src/ast/passes/clang_build.cpp

```diff
@@ -0,0 +1,212 @@
+#include "stdlib/stdlib.h"
+#include "util/result.h"
+
+namespace bpftrace::ast {
+
+char ClangBuildError::ID;
+
+void ClangBuildError::log(llvm::raw_ostream &OS) const
+{
+  OS << msg_;
+}
+
+namespace {
+
+class PipeFds {
+public:
+  PipeFds(int rfd, int wfd)
+      : rfd_(rfd),
+        wfd_(wfd),
+        read_("/dev/fd/" + std::to_string(rfd)),
+        write_("/dev/fd/" + std::to_string(wfd)) {};
+  PipeFds(PipeFds &&other)
+      : rfd_(other.rfd_),
+        wfd_(other.wfd_),
+        read_(other.read_),
+        write_(other.write_)
+  {
+    other.rfd_ = -1;
+    other.wfd_ = -1;
+  }
+  PipeFds(const PipeFds &other) = delete;
+  ~PipeFds()
+  {
+    close_read();
+    close_write();
+  }
+
+  const std::string &write_file()
+  {
+    return write_;
+  }
+
+  std::string read_all()
+  {
+    close_write();
+    std::ifstream file(read_);
+    if (file.fail()) {
+      return ""; // Nothing to read.
+    }
+    std::stringstream contents;
+    contents << file.rdbuf();
+    return contents.str();
+  }
+
+  void close_read()
+  {
+    if (rfd_ >= 0) {
+      close(rfd_);
+      rfd_ = -1;
+    }
+  }
+
+  void close_write()
+  {
+    if (wfd_ >= 0) {
+      close(wfd_);
+      wfd_ = -1;
+    }
+  }
+
+private:
+  int rfd_ = -1;
+  int wfd_ = -1;
+  std::string read_;
+  std::string write_;
+};
+
+Result<PipeFds> create_pipe()
+{
+  int fds[2];
+  if (pipe2(fds, O_CLOEXEC) < 0) {
+    return make_error<ClangBuildError>("failed to create pipes");
+  }
+  return PipeFds(fds[0], fds[1]);
+}
+
+} // namespace
+
+static Result<> build(const std::string &name,
+                      LoadedObject &obj,
+                      Imports &imports,
+                      BitcodeModules &result)
+{
+  llvm::IntrusiveRefCntPtr<llvm::vfs::InMemoryFileSystem> vfs(
+      new llvm::vfs::InMemoryFileSystem());
+  vfs->addFile(name, 0, llvm::MemoryBuffer::getMemBuffer(obj.data()));
+  for (const auto &[name, other] : stdlib::Stdlib::files) {
+    vfs->addFile(name, 0, llvm::MemoryBuffer::getMemBuffer(other));
+  }
+  for (auto &[name, other] : imports.c_headers) {
+    vfs->addFile(name, 0, llvm::MemoryBuffer::getMemBuffer(other.data()));
+  }
+
+  // Create the diagnostic options and client. We emit the error to
+  // a string, which we can then capture and associate with the import.
+  std::string errstr;
+  llvm::raw_string_ostream err(errstr);
+  auto diagOpts = llvm::makeIntrusiveRefCnt<clang::DiagnosticOptions>();
+  auto diags = std::make_unique<clang::DiagnosticsEngine>(
+      llvm::makeIntrusiveRefCnt<clang::DiagnosticIDs>(),
+      diagOpts,
+      new clang::TextDiagnosticPrinter(err, diagOpts.get()));
+
+  // We create a temporary pipe that we can use to splurp the output,
+  // since the ClangDriver API is framed in terms of filenames. Perhaps
+  // we could use the internals here, but that carries other risks.
+  auto pipefds = create_pipe();
+  if (!pipefds) {
+    return pipefds.takeError();
+  }
+
+  // Create the compiler invocation.
```

Comment:
Nit: Not a terrible place to explain why we need the `-02` or link to something that explains it.

Reply:
Done.